File: Alkitab/src/yuku/alkitab/base/ac/FontManagerActivity.java
package yuku.alkitab.base.ac;

import android.app.AlertDialog;
import android.content.ComponentName;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.ServiceConnection;
import android.os.AsyncTask;
import android.os.Bundle;
import android.os.IBinder;
import android.util.Log;
import android.view.View;
import android.view.View.OnClickListener;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.ListView;
import android.widget.ProgressBar;
import android.widget.TextView;

import java.io.BufferedInputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;
import java.util.zip.ZipEntry;
import java.util.zip.ZipInputStream;

import yuku.afw.V;
import yuku.alkitab.debug.R;
import yuku.alkitab.base.App;
import yuku.alkitab.base.ac.base.BaseActivity;
import yuku.alkitab.base.rpc.SimpleHttpConnection;
import yuku.alkitab.base.sv.DownloadService;
import yuku.alkitab.base.sv.DownloadService.DownloadBinder;
import yuku.alkitab.base.sv.DownloadService.DownloadEntry;
import yuku.alkitab.base.sv.DownloadService.DownloadListener;
import yuku.alkitab.base.util.FontManager;
import yuku.alkitab.base.widget.UrlImageView;
import yuku.alkitab.base.widget.UrlImageView.OnStateChangeListener;

public class FontManagerActivity extends BaseActivity implements DownloadListener {
	public static final String TAG = FontManagerActivity.class.getSimpleName();
	
	private static final String URL_fontList = "http://alkitab-host.appspot.com/addon/fonts/v1/list-v2.txt"; //$NON-NLS-1$
	private static final String URL_fontData = "http://alkitab-host.appspot.com/addon/fonts/v1/data/%s.zip"; //$NON-NLS-1$
	private static final String URL_fontPreview = "http://alkitab-host.appspot.com/addon/fonts/v1/preview-black/%s-384x84.png"; //$NON-NLS-1$

	public static Intent createIntent() {
		return new Intent(App.context, FontManagerActivity.class);
	}

	ListView lsFont;
	FontAdapter adapter;
	TextView lEmptyError;
	DownloadService dls;
	
	private ServiceConnection serviceConnection = new ServiceConnection() {
		@Override public void onServiceDisconnected(ComponentName name) {
			dls = null;
		}
		
		@Override public void onServiceConnected(ComponentName name, IBinder service) {
			dls = ((DownloadBinder) service).getService();
			dls.setDownloadListener(FontManagerActivity.this);
			runOnUiThread(new Runnable() {
				@Override public void run() {
					loadFontList();
				}
			});
		}
	};

	@Override protected void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);

		setContentView(R.layout.activity_font_manager);
		setTitle(R.string.fm_activity_title);

		lsFont = V.get(this, R.id.lsFont);
		lEmptyError = V.get(this, R.id.lEmptyError);
		
		lsFont.setAdapter(adapter = new FontAdapter());
		lsFont.setEmptyView(lEmptyError);
		
		bindService(new Intent(App.context, DownloadService.class), serviceConnection, BIND_AUTO_CREATE);
	}
	
	@Override protected void onDestroy() {
		super.onDestroy();
		
		if (dls != null) {
			unbindService(serviceConnection);
		}
	}

	void loadFontList() {
		new AsyncTask<Void, Void, List<FontItem>>() {
			String errorMsg;
			
			@Override protected List<FontItem> doInBackground(Void... params) {
				SimpleHttpConnection conn = new SimpleHttpConnection(URL_fontList);
				try {
					InputStream in = conn.load();
					if (in == null) {
						Exception ex = conn.getException();
						errorMsg = ex.getClass().getSimpleName() + " " + ex.getMessage(); //$NON-NLS-1$
						return null;
					} else {
						if (conn.isSameHost()) {
							List<FontItem> list = new ArrayList<FontItem>();
							Scanner sc = new Scanner(in);
							if (sc.hasNextLine() && sc.nextLine().startsWith("OK")) { //$NON-NLS-1$
								while (sc.hasNextLine()) {
									String line = sc.nextLine().trim();
									if (line.length() > 0) {
										FontItem item = new FontItem();
										item.name = line.split(" ")[0]; //$NON-NLS-1$
										list.add(item);
									}
								}
							}
							return list;
						} else {
							errorMsg = getString(R.string.fm_need_to_log_in_to_wifi);
						}
						return null;
					}
				} finally {
					conn.close();
				}
			}
			
			@Override protected void onPostExecute(List<FontItem> result) {
				if (result != null) {
					lEmptyError.setText(null);
					adapter.setData(result);
				} else {
					lEmptyError.setText(errorMsg);
				}
			};
		}.execute();
	}
	
	String getFontDownloadKey(String name) {
		return "FontManager/" + name; //$NON-NLS-1$
	}
	
	private String getFontNameFromDownloadKey(String key) {
		if (!key.startsWith("FontManager/")) return null; //$NON-NLS-1$
		return key.substring("FontManager/".length()); //$NON-NLS-1$
	}

	String getFontDownloadDestination(String name) {
		return new File(FontManager.getFontsPath(), "download-" + name + ".zip").getAbsolutePath(); //$NON-NLS-1$ //$NON-NLS-2$
	}
	
	public static class FontItem {
		public String name;
	}
	
	public class FontAdapter extends BaseAdapter {
		List<FontItem> list;
		
		public void setData(List<FontItem> list) {
			this.list = list;
			notifyDataSetChanged();
		}

		@Override public int getCount() {
			return list == null? 0: list.size();
		}

		@Override public FontItem getItem(int position) {
			return list.get(position);
		}

		@Override public long getItemId(int position) {
			return position;
		}

		@Override public View getView(int position, View convertView, ViewGroup parent) {
			View res = convertView != null ? convertView : getLayoutInflater().inflate(R.layout.item_font_download, null);

			UrlImageView imgPreview = V.get(res, R.id.imgPreview);
			TextView lFontName = V.get(res, R.id.lFontName);
			View bDownload = V.get(res, R.id.bDownload);
			View bDelete = V.get(res, R.id.bDelete);
			ProgressBar progressbar = V.get(res, R.id.progressbar);
			TextView lErrorMsg = V.get(res, R.id.lErrorMsg);
			
			FontItem item = getItem(position);
			String dlkey = getFontDownloadKey(item.name);
			
			lFontName.setText(item.name);
			lFontName.setVisibility(View.VISIBLE);
			imgPreview.setTag(R.id.TAG_fontName, lFontName);
			imgPreview.setOnStateChangeListener(imgPreview_stateChange);
			imgPreview.setUrl(String.format(URL_fontPreview, item.name));
			bDownload.setTag(R.id.TAG_fontItem, item);
			bDownload.setOnClickListener(bDownload_click);
			bDelete.setTag(R.id.TAG_fontItem, item);
			bDelete.setOnClickListener(bDelete_click);
			
			if (FontManager.isInstalled(item.name)) {
				progressbar.setIndeterminate(false);
				progressbar.setMax(100);
				progressbar.setProgress(100);
				bDownload.setVisibility(View.GONE);
				bDelete.setVisibility(View.VISIBLE);
				lErrorMsg.setVisibility(View.GONE);
			} else {
				DownloadEntry entry = dls.getEntry(dlkey);
				if (entry == null) {
					progressbar.setIndeterminate(false);
					progressbar.setMax(100);
					progressbar.setProgress(0);
					bDownload.setVisibility(View.VISIBLE);
					bDownload.setEnabled(true);
					bDelete.setVisibility(View.GONE);
					lErrorMsg.setVisibility(View.GONE);
				} else {
					if (entry.state == DownloadService.State.created) {
						progressbar.setIndeterminate(true);
						bDownload.setVisibility(View.VISIBLE);
						bDownload.setEnabled(false);
						bDelete.setVisibility(View.GONE);
						lErrorMsg.setVisibility(View.GONE);
					} else if (entry.state == DownloadService.State.downloading) {
						if (entry.length == -1) {
							progressbar.setIndeterminate(true);
						} else {
							progressbar.setIndeterminate(false);
							progressbar.setMax((int) entry.length);
							progressbar.setProgress((int) entry.progress);
						}
						bDownload.setVisibility(View.VISIBLE);
						bDownload.setEnabled(false);
						bDelete.setVisibility(View.GONE);
						lErrorMsg.setVisibility(View.GONE);
					} else if (entry.state == DownloadService.State.finished) {
						progressbar.setIndeterminate(false);
						progressbar.setMax(100); // consider full
						progressbar.setProgress(100); // consider full
						bDownload.setVisibility(View.GONE);
						bDelete.setVisibility(View.VISIBLE);
						lErrorMsg.setVisibility(View.GONE);
					} else if (entry.state == DownloadService.State.failed) {
						progressbar.setIndeterminate(false);
						progressbar.setMax(100);
						progressbar.setProgress(0);
						bDownload.setVisibility(View.VISIBLE);
						bDownload.setEnabled(true);
						bDelete.setVisibility(View.GONE);
						lErrorMsg.setVisibility(View.VISIBLE);
						lErrorMsg.setText(entry.errorMsg);
					}
				}
			}
			
			return res;
		}
		
		private OnClickListener bDownload_click = new OnClickListener() {
			@Override public void onClick(View v) {
				FontItem item = (FontItem) v.getTag(R.id.TAG_fontItem);
				
				String dlkey = getFontDownloadKey(item.name);
				dls.removeEntry(dlkey);
				
				if (dls.getEntry(dlkey) == null) {
					new File(FontManager.getFontsPath()).mkdirs();
					dls.startDownload(
						dlkey,
						String.format(URL_fontData, item.name),
						getFontDownloadDestination(item.name)
					);
				}
				
				notifyDataSetChanged();
			}
		};
		
		private OnClickListener bDelete_click = new OnClickListener() {
			@Override public void onClick(View v) {
				final FontItem item = (FontItem) v.getTag(R.id.TAG_fontItem);
				
				new AlertDialog.Builder(FontManagerActivity.this)
				.setMessage(getString(R.string.fm_do_you_want_to_delete, item.name))
				.setPositiveButton(R.string.yes, new DialogInterface.OnClickListener() {
					@Override public void onClick(DialogInterface dialog, int which) {
						File fontDir = FontManager.getFontDir(item.name);
						File[] listFiles = fontDir.listFiles();
						if (listFiles != null) {
							for (File file: listFiles) {
								file.delete();
							}
						}
						fontDir.delete();
						
						dls.removeEntry(getFontDownloadKey(item.name));
						
						notifyDataSetChanged();
					}
				})
				.setNegativeButton(R.string.no, null)
				.show();
			}
		};
		
		private OnStateChangeListener imgPreview_stateChange = new OnStateChangeListener() {
			@Override public void onStateChange(UrlImageView v, UrlImageView.State newState, String url) {
				if (newState.isLoaded()) {
					TextView lFontName = (TextView) v.getTag(R.id.TAG_fontName);
					lFontName.setVisibility(View.GONE);
				}
			}
		};
	}

	@Override public void onStateChanged(DownloadEntry entry, DownloadService.State originalState) {
		adapter.notifyDataSetChanged();
		
		if (originalState == DownloadService.State.finished) {
			String fontName = getFontNameFromDownloadKey(entry.key);
			if (fontName == null) { // this download doesn't belong to font manager.
				return;
			}
			
			try {
				String downloadedZip = getFontDownloadDestination(fontName);
				File fontDir = FontManager.getFontDir(fontName);
				fontDir.mkdirs();
				
				Log.d(TAG, "Going to unzip " + downloadedZip, new Throwable().fillInStackTrace()); //$NON-NLS-1$
				
				ZipInputStream zis = new ZipInputStream(new BufferedInputStream(new FileInputStream(downloadedZip)));
				try {
					ZipEntry ze;
					while ((ze = zis.getNextEntry()) != null) {
						String zname = ze.getName();
						Log.d(TAG, "Extracting from zip: " + zname); //$NON-NLS-1$
						File extractFile = new File(fontDir, zname);
						FileOutputStream fos = new FileOutputStream(extractFile);
						try {
							byte[] buf = new byte[4096];
							int count;
							while ((count = zis.read(buf)) != -1) {
								fos.write(buf, 0, count);
							}
						} finally {
							fos.close();
						}
					}
				} finally {
					zis.close();
				}
				
				new File(downloadedZip).delete();
			} catch (Exception e) {
				new AlertDialog.Builder(FontManagerActivity.this)
				.setMessage(getString(R.string.fm_error_when_extracting_font, fontName, e.getClass().getSimpleName() + ' ' + e.getMessage()))
				.setPositiveButton(R.string.ok, null)
				.show();
			}
		}
	}

	@Override public void onProgress(DownloadEntry entry, DownloadService.State originalState) {
		adapter.notifyDataSetChanged();
	}
}

File: Alkitab/src/main/java/yuku/alkitab/base/ac/FontManagerActivity.kt
package yuku.alkitab.base.ac

import android.content.ComponentName
import android.content.Intent
import android.content.ServiceConnection
import android.os.Bundle
import android.os.IBinder
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.ListView
import android.widget.ProgressBar
import android.widget.TextView
import com.afollestad.materialdialogs.MaterialDialog
import com.jakewharton.picasso.OkHttp3Downloader
import com.squareup.picasso.Callback.EmptyCallback
import com.squareup.picasso.Picasso
import java.io.BufferedInputStream
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.io.IOException
import java.util.zip.ZipEntry
import java.util.zip.ZipInputStream
import yuku.afw.widget.EasyAdapter
import yuku.alkitab.base.App
import yuku.alkitab.base.ac.base.BaseActivity
import yuku.alkitab.base.connection.Connections
import yuku.alkitab.base.sv.DownloadService
import yuku.alkitab.base.util.AppLog
import yuku.alkitab.base.util.Background
import yuku.alkitab.base.util.FontManager
import yuku.alkitab.base.util.Foreground
import yuku.alkitab.debug.BuildConfig
import yuku.alkitab.debug.R

private const val TAG = "FontManagerActivity"
private const val URL_fontList = BuildConfig.SERVER_HOST + "addon/fonts/v1/list-v2.txt"
private const val URL_fontData = BuildConfig.SERVER_HOST + "addon/fonts/v1/data/%s.zip"
private const val URL_fontPreview = BuildConfig.SERVER_HOST + "addon/fonts/v1/preview/%s-384x84.png"

class FontManagerActivity : BaseActivity(), DownloadService.DownloadListener {
    private lateinit var lsFont: ListView
    private lateinit var adapter: FontAdapter
    private lateinit var progress: View
    private lateinit var lEmptyError: TextView

    private var dls: DownloadService? = null

    private val picasso: Picasso by lazy {
        Picasso.Builder(this)
            .downloader(OkHttp3Downloader(Connections.okHttp))
            .build()
    }

    private val serviceConnection: ServiceConnection = object : ServiceConnection {
        override fun onServiceDisconnected(name: ComponentName) {
            dls = null
        }

        override fun onServiceConnected(name: ComponentName, service: IBinder) {
            val dls = (service as DownloadService.DownloadBinder).service
            dls.setDownloadListener(this@FontManagerActivity)
            this@FontManagerActivity.dls = dls
            runOnUiThread { loadFontList() }
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.willNeedStoragePermission()
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_font_manager)

        setSupportActionBar(findViewById(R.id.toolbar))
        supportActionBar?.setDisplayHomeAsUpEnabled(true)

        lsFont = findViewById(R.id.lsFont)
        progress = findViewById(R.id.progress)
        lEmptyError = findViewById(R.id.lEmptyError)
        adapter = FontAdapter()
        lsFont.adapter = adapter

        bindService(Intent(this, DownloadService::class.java), serviceConnection, BIND_AUTO_CREATE)
    }

    override fun onNeededPermissionsGranted(immediatelyGranted: Boolean) {
        super.onNeededPermissionsGranted(immediatelyGranted)
        if (!immediatelyGranted && dls != null) {
            loadFontList()
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        if (dls != null) {
            unbindService(serviceConnection)
        }
    }

    fun loadFontList() = Background.run {
        try {
            val listString = App.downloadString(URL_fontList)
            val list = mutableListOf<FontItem>()
            val lines = listString.lines()
            if (lines.firstOrNull() == "OK") {
                for (line in lines.drop(1)) {
                    if (line.isNotBlank()) {
                        list.add(FontItem(line.trim().split(" ")[0]))
                    }
                }
            }
            Foreground.run {
                adapter.setData(list)
                lEmptyError.visibility = View.GONE
                progress.visibility = View.GONE
            }
        } catch (e: IOException) {
            Foreground.run {
                lEmptyError.visibility = View.VISIBLE
                lEmptyError.text = e.message
                progress.visibility = View.GONE
            }
        }
    }

    fun getFontDownloadKey(name: String?): String {
        return "FontManager/$name"
    }

    private fun getFontNameFromDownloadKey(key: String): String? {
        return if (!key.startsWith("FontManager/")) null else key.substring("FontManager/".length)
    }

    fun getFontDownloadDestination(name: String?): String {
        return File(cacheDir, "download-$name.zip").absolutePath
    }

    data class FontItem(val name: String)

    inner class FontAdapter : EasyAdapter() {
        private val list = mutableListOf<FontItem>()

        fun setData(list: List<FontItem>) {
            this.list.clear()
            this.list.addAll(list)
            notifyDataSetChanged()
        }

        override fun getCount(): Int {
            return list.size
        }

        override fun getItem(position: Int): FontItem {
            return list[position]
        }

        override fun newView(position: Int, parent: ViewGroup): View {
            return layoutInflater.inflate(R.layout.item_font_download, parent, false)
        }

        override fun bindView(view: View, position: Int, parent: ViewGroup) {
            val imgPreview = view.findViewById<ImageView>(R.id.imgPreview)
            val lFontName = view.findViewById<TextView>(R.id.lFontName)
            val bDownload = view.findViewById<View>(R.id.bDownload)
            val bDelete = view.findViewById<View>(R.id.bDelete)
            val progressbar = view.findViewById<ProgressBar>(R.id.progressbar)
            val lErrorMsg = view.findViewById<TextView>(R.id.lErrorMsg)

            val item = getItem(position)
            val dlkey = getFontDownloadKey(item.name)

            lFontName.text = item.name
            lFontName.visibility = View.VISIBLE

            picasso.load(String.format(URL_fontPreview, item.name)).into(imgPreview, object : EmptyCallback() {
                override fun onError(e: Exception) {
                    AppLog.e(TAG, "error loading font preview", e)
                }

                override fun onSuccess() {
                    lFontName.visibility = View.GONE
                }
            })

            imgPreview.contentDescription = item.name
            bDownload.setTag(R.id.TAG_fontItem, item)
            bDownload.setOnClickListener(bDownload_click)
            bDelete.setTag(R.id.TAG_fontItem, item)
            bDelete.setOnClickListener(bDelete_click)
            if (FontManager.isInstalled(item.name)) {
                progressbar.isIndeterminate = false
                progressbar.max = 100
                progressbar.progress = 100
                bDownload.visibility = View.GONE
                bDelete.visibility = View.VISIBLE
                lErrorMsg.visibility = View.GONE
                return
            }

            val entry = dls?.getEntry(dlkey)
            if (entry == null) {
                progressbar.isIndeterminate = false
                progressbar.max = 100
                progressbar.progress = 0
                bDownload.visibility = View.VISIBLE
                bDownload.isEnabled = true
                bDelete.visibility = View.GONE
                lErrorMsg.visibility = View.GONE
                return
            }

            when (entry.state) {
                DownloadService.State.created -> {
                    progressbar.isIndeterminate = true
                    bDownload.visibility = View.VISIBLE
                    bDownload.isEnabled = false
                    bDelete.visibility = View.GONE
                    lErrorMsg.visibility = View.GONE
                }

                DownloadService.State.downloading -> {
                    if (entry.length == -1L) {
                        progressbar.isIndeterminate = true
                    } else {
                        progressbar.isIndeterminate = false
                        progressbar.max = entry.length.toInt()
                        progressbar.progress = entry.progress.toInt()
                    }
                    bDownload.visibility = View.VISIBLE
                    bDownload.isEnabled = false
                    bDelete.visibility = View.GONE
                    lErrorMsg.visibility = View.GONE
                }

                DownloadService.State.finished -> {
                    progressbar.isIndeterminate = false
                    progressbar.max = 100 // consider full
                    progressbar.progress = 100 // consider full
                    bDownload.visibility = View.GONE
                    bDelete.visibility = View.VISIBLE
                    lErrorMsg.visibility = View.GONE
                }

                DownloadService.State.failed -> {
                    progressbar.isIndeterminate = false
                    progressbar.max = 100
                    progressbar.progress = 0
                    bDownload.visibility = View.VISIBLE
                    bDownload.isEnabled = true
                    bDelete.visibility = View.GONE
                    lErrorMsg.visibility = View.VISIBLE
                    lErrorMsg.text = entry.errorMsg
                }

                else -> {
                }
            }
        }

        private val bDownload_click = View.OnClickListener { v ->
            val dls = dls ?: return@OnClickListener
            val item = v.getTag(R.id.TAG_fontItem) as FontItem
            val dlkey = getFontDownloadKey(item.name)
            dls.removeEntry(dlkey)
            if (dls.getEntry(dlkey) == null) {
                dls.startDownload(dlkey, String.format(URL_fontData, item.name), getFontDownloadDestination(item.name))
            }
            notifyDataSetChanged()
        }

        private val bDelete_click = View.OnClickListener { v ->
            val dls = dls ?: return@OnClickListener
            val item = v.getTag(R.id.TAG_fontItem) as FontItem
            MaterialDialog.Builder(this@FontManagerActivity)
                .content(getString(R.string.fm_do_you_want_to_delete, item.name))
                .positiveText(R.string.delete)
                .onPositive { _, _ ->
                    val fontDir = FontManager.getFontDir(item.name)
                    val listFiles = fontDir.listFiles()
                    if (listFiles != null) {
                        for (file in listFiles) {
                            file.delete()
                        }
                    }
                    fontDir.delete()
                    dls.removeEntry(getFontDownloadKey(item.name))
                    notifyDataSetChanged()
                }
                .negativeText(R.string.cancel)
                .show()
        }
    }

    override fun onStateChanged(entry: DownloadService.DownloadEntry, originalState: DownloadService.State) {
        adapter.notifyDataSetChanged()

        if (originalState != DownloadService.State.finished) return
        val fontName = getFontNameFromDownloadKey(entry.key) ?: return // this download doesn't belong to font manager.

        try {
            val downloadedZip = getFontDownloadDestination(fontName)
            val fontDir = FontManager.getFontDir(fontName)
            fontDir.mkdirs()

            AppLog.d(TAG, "Going to unzip $downloadedZip", Throwable().fillInStackTrace())

            ZipInputStream(BufferedInputStream(FileInputStream(downloadedZip))).use { zis ->
                while (true) {
                    val ze: ZipEntry = zis.nextEntry ?: break
                    val zname = ze.name
                    AppLog.d(TAG, "Extracting from zip: $zname")

                    val extractFile = File(fontDir, zname)

                    // https://support.google.com/faqs/answer/9294009
                    val fontDirPath = fontDir.canonicalPath
                    val extractFilePath = extractFile.canonicalPath
                    if (!extractFilePath.startsWith(fontDirPath)) {
                        throw SecurityException("Zip path traversal attack: $fontDirPath, $zname")
                    }

                    FileOutputStream(extractFile).use { fos -> zis.copyTo(fos) }
                }
            }
            File(downloadedZip).delete()

        } catch (e: Exception) {
            MaterialDialog.Builder(this@FontManagerActivity)
                .content(getString(R.string.fm_error_when_extracting_font, fontName, "$e"))
                .positiveText(R.string.ok)
                .show()
        }
    }

    override fun onProgress(entry: DownloadService.DownloadEntry, originalState: DownloadService.State) {
        adapter.notifyDataSetChanged()
    }

    companion object {
        @JvmStatic
        fun createIntent(): Intent {
            return Intent(App.context, FontManagerActivity::class.java)
        }
    }
}
